### PR TITLE
[Bastion] Add support for Bastion shareable links

### PR DESCRIPTION
**Pull request by abhi7860**
## Mandatory Checklist

- Please choose the target release of Azure PowerShell. (⚠️**Target release** is a different concept from **API readiness**. Please click below links for details.)
  - [x] [General release](../blob/main/CONTRIBUTING.md#target-release-types)
  - [ ] [Public preview](../blob/main/CONTRIBUTING.md#target-release-types)
  - [ ] [Private preview](../blob/main/CONTRIBUTING.md#target-release-types)
  - [ ] [Engineering build](../blob/main/CONTRIBUTING.md#target-release-types)
  - [ ] No need for a release

- [x] Check this box to confirm: **I have read the [_Submitting Changes_](../blob/main/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](../blob/main/CONTRIBUTING.md) and reviewed the following information:**

* **SHOULD** update `ChangeLog.md` file(s) appropriately
    * For SDK-based development mode, update `src/{{SERVICE}}/{{SERVICE}}/ChangeLog.md`.
        * A snippet outlining the change(s) made in the PR should be written under the `## Upcoming Release` header in the past tense. 
    * For autorest-based development mode, include the changelog in the PR description.
    * Should **not** change `ChangeLog.md` if no new release is required, such as fixing test case only.
* **SHOULD** regenerate markdown help files if there is cmdlet API change. [Instruction](../blob/main/documentation/development-docs/help-generation.md#updating-all-markdown-files-in-a-module)
* **SHOULD** have proper test coverage for changes in pull request.
* **SHOULD NOT** adjust version of module manually in pull request

Issue: https://github.com/Azure/azure-powershell-cmdlet-review-pr/issues/1428

### CODE REVIEW

**Comment by VeryEarly** [comment]
Please move changelog to line 22 right under 'upcoming release'

**Review by VeryEarly — DISMISSED** [review]

**Comment by VeryEarly** [comment]
please fix below example issues:
"Az.Network","Get-AzBastionShareableLink","1","0","NeedDeleting","5050","1","The prompt of example need to be deleted.","D:\a\_work\1\s\src\Network\Network\help\Get-AzBastionShareableLink.md","Delete the prompt of example."
  "Az.Network","Get-AzBastionShareableLink","2","0","NeedDeleting","5050","1","The prompt of example need to be deleted.","D:\a\_work\1\s\src\Network\Network\help\Get-AzBastionShareableLink.md","Delete the prompt of example."
  "Az.Network","New-AzBastionShareableLink","1","0","NeedDeleting","5050","1","The prompt of example need to be deleted.","D:\a\_work\1\s\src\Network\Network\help\New-AzBastionShareableLink.md","Delete the prompt of example."
  "Az.Network","New-AzBastionShareableLink","2","0","NeedDeleting","5050","1","The prompt of example need to be deleted.","D:\a\_work\1\s\src\Network\Network\help\New-AzBastionShareableLink.md","Delete the prompt of example."
  "Az.Network","Remove-AzBastionShareableLink","1","0","NeedDeleting","5050","1","The prompt of example need to be deleted.","D:\a\_work\1\s\src\Network\Network\help\Remove-AzBastionShareableLink.md","Delete the prompt of example."
  "Az.Network","Remove-AzBastionShareableLink","2","0","NeedDeleting","5050","1","The prompt of example need to be deleted.","D:\a\_work\1\s\src\Network\Network\help\Remove-AzBastionShareableLink.md","Delete the prompt of example."
  "Az.Network","Remove-AzBastionShareableLink","1","4","UnexpectedToken","5200","1","Unexpected token 'Yes' in expression or statement.","Yes","Unexpected Error. Please check [Trouble Shotting for Unexpected Errors in Example Issues](https://github.com/Azure/azure-powershell/blob/main/documentation/Debugging-StaticAnalysis-Errors.md#Troubleshotting-Example-Issues) for more details."
  "Az.Network","Remove-AzBastionShareableLink","1","2","Invalid_Cmdlet","5000","1","Confirm is not a valid command name.","Confirm","Check the spell of Confirm."
  "Az.Network","Remove-AzBastionShareableLink","1","3","Invalid_Cmdlet","5000","1","Are is not a valid command name.","Are you sure you want to remove Bastion shareable link(s)?","Check the spell of Are."
  "Az.Network","Remove-AzBastionShareableLink","1","3","Invalid_Cmdlet","5000","1","s is not a valid command name.","s","Check the spell of s."
  "Az.Network","Remove-AzBastionShareableLink","1","4","Capitalization_Conventions_Violated","5101","1","Yes doesn't follow the Capitalization Conventions.","Yes  [N] No  [S] Suspend  [?] Help (default is 'Y'): Y","Check the Capitalization Conventions. Suggest format: yes.exe"
  "Az.Network","Remove-AzBastionShareableLink","1","4","Invalid_Cmdlet","5000","1","default is not a valid command name.","default is 'Y'","Check the spell of default."

File: src/Network/Network/help/New-AzBastionShareableLink.md [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+  "createdAt": "2019-10-18T12:00:00.0000Z"
+}
+```
+
+### Example 2
+```powershell
+PS C:\> $vm1 = Get-AzVM -ResourceGroupName $RgName -Name $vmName1
+PS C:\> $vm2 = Get-AzVM -ResourceGroupName $RgName -Name $vmName2
+PS C:\> $bastion = Get-AzBastion -ResourceGroupName $RgName -Name $bastionName
+PS C:\> New-AzBastionShareableLink -InputObject $bastion -TargetVmId $vm1.Id, $vm2.Id
```
Comment (msftsijielin) [inline src/Network/Network/help/New-AzBastionShareableLink.md:64]: wonder if there can be space between target vm ids

**Review by msftsijielin — APPROVED** [review]

**Comment by msJinLei** [comment]
```
  Analyzer invoked with parameters: -p D:\a\_work\1\s\artifacts/Debug -r D:\a\_work\1\s\artifacts/StaticAnalysisResults --analyzers check-error
  D:\a\_work\1\s\artifacts/StaticAnalysisResults
  D:\a\_work\1\s\artifacts/StaticAnalysisResults\ExampleIssues.csv Errors
  "Module","Cmdlet","Example","Line","RuleName","ProblemId","Severity","Description","Extent","Remediation"
  "Az.Network","Get-AzBastionShareableLink","1","0","NeedDeleting","5050","1","The prompt of example need to be deleted.","D:\a\_work\1\s\src\Network\Network\help\Get-AzBastionShareableLink.md","Delete the prompt of example."
  "Az.Network","Get-AzBastionShareableLink","2","0","NeedDeleting","5050","1","The prompt of example need to be deleted.","D:\a\_work\1\s\src\Network\Network\help\Get-AzBastionShareableLink.md","Delete the prompt of example."
  "Az.Network","New-AzBastionShareableLink","1","0","NeedDeleting","5050","1","The prompt of example need to be deleted.","D:\a\_work\1\s\src\Network\Network\help\New-AzBastionShareableLink.md","Delete the prompt of example."
  "Az.Network","New-AzBastionShareableLink","2","0","NeedDeleting","5050","1","The prompt of example need to be deleted.","D:\a\_work\1\s\src\Network\Network\help\New-AzBastionShareableLink.md","Delete the prompt of example."
  "Az.Network","Remove-AzBastionShareableLink","1","0","NeedDeleting","5050","1","The prompt of example need to be deleted.","D:\a\_work\1\s\src\Network\Network\help\Remove-AzBastionShareableLink.md","Delete the prompt of example."
  "Az.Network","Remove-AzBastionShareableLink","2","0","NeedDeleting","5050","1","The prompt of example need to be deleted.","D:\a\_work\1\s\src\Network\Network\help\Remove-AzBastionShareableLink.md","Delete the prompt of example."
  "Az.Network","Remove-AzBastionShareableLink","1","4","UnexpectedToken","5200","1","Unexpected token 'Yes' in expression or statement.","Yes","Unexpected Error. Please check [Trouble Shotting for Unexpected Errors in Example Issues](https://github.com/Azure/azure-powershell/blob/main/documentation/Debugging-StaticAnalysis-Errors.md#Troubleshotting-Example-Issues) for more details."
  "Az.Network","Remove-AzBastionShareableLink","1","2","Invalid_Cmdlet","5000","1","Confirm is not a valid command name.","Confirm","Check the spell of Confirm."
  "Az.Network","Remove-AzBastionShareableLink","1","3","Invalid_Cmdlet","5000","1","Are is not a valid command name.","Are you sure you want to remove Bastion shareable link(s)?","Check the spell of Are."
  "Az.Network","Remove-AzBastionShareableLink","1","3","Invalid_Cmdlet","5000","1","s is not a valid command name.","s","Check the spell of s."
  "Az.Network","Remove-AzBastionShareableLink","1","4","Capitalization_Conventions_Violated","5101","1","Yes doesn't follow the Capitalization Conventions.","Yes  [N] No  [S] Suspend  [?] Help (default is 'Y'): Y","Check the Capitalization Conventions. Suggest format: yes.exe"
  "Az.Network","Remove-AzBastionShareableLink","1","4","Invalid_Cmdlet","5000","1","default is not a valid command name.","default is 'Y'","Check the spell of default."
```

File: src/Network/Network/help/Remove-AzBastionShareableLink.md [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+
+## DESCRIPTION
+Deletes the Bastion Shareable Links for all the VMs specified in the request.
+
+## EXAMPLES
+
+### Example 1
+```powershell
+PS C:\> Remove-AzBastionShareableLink -ResourceGroupName $RgName -Name $bastionName -TargetVmId $vm.Id
+Confirm
```
Comment (msJinLei) [inline src/Network/Network/help/Remove-AzBastionShareableLink.md:47]: Please remove non-powershell text

File: src/Network/Network/help/New-AzBastionShareableLink.md [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+```
+
+## DESCRIPTION
+Creates a Bastion Shareable Links for all the VMs specified in the request.
+
+## EXAMPLES
+
+### Example 1
+```powershell
+PS C:\> $vm = Get-AzVM -ResourceGroupName $RgName -Name $vmName
```
Comment (msJinLei) [inline src/Network/Network/help/New-AzBastionShareableLink.md:45]:
```suggestion
$vm = Get-AzVM -ResourceGroupName $RgName -Name $vmName
```

File: src/Network/Network/help/New-AzBastionShareableLink.md [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+PS C:\> New-AzBastionShareableLink -ResourceGroupName $RgName -Name $bastionName -TargetVmId $vm.Id
```
Comment (msJinLei) [inline src/Network/Network/help/New-AzBastionShareableLink.md:46]:
```suggestion
New-AzBastionShareableLink -ResourceGroupName $RgName -Name $bastionName -TargetVmId $vm.Id
```

File: src/Network/Network/help/New-AzBastionShareableLink.md [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+    "id": "/subscriptions/subid/resourceGroups/rgx/providers/Microsoft.Compute/virtualMachines/vm1"
+  },
+  "bsl": "http://bst-bastionhostid.bastion.com/api/shareable-url/tokenvm1",
+  "createdAt": "2019-10-18T12:00:00.0000Z"
+}
+```
+
+### Example 2
+```powershell
+PS C:\> $vm1 = Get-AzVM -ResourceGroupName $RgName -Name $vmName1
```
Comment (msJinLei) [inline src/Network/Network/help/New-AzBastionShareableLink.md:61]: $vm1 = Get-AzVM -ResourceGroupName $RgName -Name $vmName1
$vm2 = Get-AzVM -ResourceGroupName $RgName -Name $vmName2
$bastion = Get-AzBastion -ResourceGroupName $RgName -Name $bastionName
New-AzBastionShareableLink -InputObject $bastion -TargetVmId $vm1.Id, $vm2.Id

File: src/Network/Network/help/Remove-AzBastionShareableLink.md [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+```
+
+## DESCRIPTION
+Deletes the Bastion Shareable Links for all the VMs specified in the request.
+
+## EXAMPLES
+
+### Example 1
+```powershell
+PS C:\> Remove-AzBastionShareableLink -ResourceGroupName $RgName -Name $bastionName -TargetVmId $vm.Id
```
Comment (msJinLei) [inline src/Network/Network/help/Remove-AzBastionShareableLink.md:46]:
```suggestion
Remove-AzBastionShareableLink -ResourceGroupName $RgName -Name $bastionName -TargetVmId $vm.Id
```

File: src/Network/Network/help/Remove-AzBastionShareableLink.md [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+Confirm
+Are you sure you want to remove Bastion shareable link(s)?
+[Y] Yes  [N] No  [S] Suspend  [?] Help (default is "Y"): Y
+```
+
+### Example 2
+```powershell
+PS C:\> Remove-AzBastionShareableLink -InputObject $bastion -TargetVmId $vm.Id -Force
```
Comment (msJinLei) [inline src/Network/Network/help/Remove-AzBastionShareableLink.md:54]:
```suggestion
Remove-AzBastionShareableLink -InputObject $bastion -TargetVmId $vm.Id -Force
```

**Review by msJinLei — CHANGES_REQUESTED** [review]
Please fix the help doc issues

File: src/Network/Network.Test/ScenarioTests/BastionTests.ps1 [hunk 8]
```diff
@@ -462,3 +462,103 @@ function Test-BastionIpObjectParam {
...
+<#
+Test Bastion Shareable Link
+#>
+function Test-BastionShareableLink {
+    # Setup
+    $rgname = Get-ResourceGroupName
+    $resourceTypeParent = "Microsoft.Network/bastionHosts"
+    $location = Get-ProviderLocation $resourceTypeParent
+    $subnetName = "AzureBastionSubnet"
+    $vnetName = "$(Get-ResourceName)-Vnet"
```
[the remark this answers is not in the document]
Comment (abhi7860) [reply src/Network/Network.Test/ScenarioTests/BastionTests.ps1:475]: "Get-ResourceName" is a function that generates and returns a new random string for resource names every time

File: src/Network/Network/Bastion/ShareableLink/GetAzBastionShareableLinkCommand.cs [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+            {
+                throw new ItemNotFoundException(string.Format(Properties.Resources.ResourceNotFound, this.Name));
+            }
+
+            if (!bastion.EnableShareableLink.Value)
+            {
+                throw new PropertyNotFoundException(Properties.Resources.BastionShareableLinkNotEnabled);
+            }
+
+            var psBslRequest = new PSBastionShareableLinkRequest(this.TargetVmId);
```
[the remark this answers is not in the document]
Comment (abhi7860) [reply src/Network/Network/Bastion/ShareableLink/GetAzBastionShareableLinkCommand.cs:102]: Yes

File: src/Network/Network/help/New-AzBastionShareableLink.md (hunk 1 above)
Comment (abhi7860) [reply src/Network/Network/help/New-AzBastionShareableLink.md]: Should not block

File: src/Network/Network/Network.format.ps1xml [hunk 10]
```diff
@@ -7088,6 +7088,84 @@
...
+      <ListControl>
+        <ListEntries>
+          <ListEntry>
+            <ListItems>
+              <ListItem>
+                <Label>VM Resource ID</Label>
+                <PropertyName>VmId</PropertyName>
+              </ListItem>
+              <ListItem>
+                <Label>BSL</Label>
```
[the remark this answers is not in the document]
Comment (abhi7860) [reply src/Network/Network/Network.format.ps1xml]: Updated the label

**Comment by abhi7860** [comment]
> Please fix the help doc issues

 @msJinLei Issues with help docs have been addressed. Can you please review again? Thanks!

**Review by msJinLei — DISMISSED** [review]

**Review by msJinLei — APPROVED** [review]

**Pull request merged**